### ISSUE TITLE
Test discoverer reports "cannot find path" errors for large test files with many ajax calls

### ISSUE THREAD

**Issue by seans83** [mmanela/chutzpah#552]
We have a large project with many test files. These test files test methods and objects that make use of many $.getJSON and $.ajax calls. When a test file grows to a certain size, these ajax calls start to report errors in the output window e.g.:

Error: Error: Error opening /Controller/SomeMethodName: The system cannot find the path specified.
While Running:C:\test-file.ts

This is reported during test discovery after making a change to a test file.

We've found that by breaking these large test files down into smaller test files, this temporarily fixes the problem, however as those test files grow, the error returns.

This error doesn't prevent Visual Studio Test Explorer from being able to run the tests, however, since they are reported on TFS as well, they do prevent the Test build step from completing.

NB: We are using the jasmine testing framework.

**Comment by mmanela** [comment]
Can you provide me a small repro of this that I can test?

**Comment by seans83** [comment]
I can't provide you with the actual repo of this, unfortunately. If I have some time, I might be able to put together an example that returns the same error.

Essentially, it would be an object that has many ajax/getJSON calls, and a few tests on that object. I've found that once you get around 15 tests in the file - the error starts to occur.

**Comment by seans83** [comment]
Ok, I've created a test project. Am I able to email to you?

**Comment by mmanela** [comment]
Easiest is to create a git repo and push to github

**Comment by seans83** [comment]
https://github.com/seans83/chutzpahtest

**Comment by seans83** [comment]
Incidentally, I've found that if I remove the test file path from Includes in chutzpah.json, the tests are discovered but the error does not occur. Of course, the tests themselves fail in this case because they can't find object they are testing.

**Comment by seans83** [comment]
Have you had a chance to look at this one yet? Is there anything more info you need?

**Comment by mmanela** [comment]
Thanks I dug into it a bit and a couple things
1. Why is your code making a Ajax request for a resource that does not exist? That does not look right to me for unit tests.
2. In the current version of Chutzpah it uses the local filesystem and it can't find this file (since it doesnt exist and errors)
3. In the next version of Chutzpah I support running as a web server and it still errors but now it seems because of a PhantomJS but (https://github.com/ariya/phantomjs/issues/12750)


But I recommend you revisit what your test is doing since it seems wrong.

**Comment by seans83** [comment]
Ok, if you take half of the tests out of that test file, you'll find the error doesn't get reported. So, it's got nothing to do with what the actual request is. At some point, Chutzpah decides it needs to treat the request url as a path that for some reason it needs to locate - it will never be able to locate the path since the project is not running while the unit tests are being performed. 

The example project I provided is just a way to replicate the error - but basically, the test engine should still be able to discover that file without reporting any errors. The number of requests and their location or availability should be irrelevant.

**Comment by mmanela** [comment]
Thanks @seans83 for the feedback. I was able to reproduce this issue but have no idea why it is happening. Chutzpah has little to do with this, it seems to be some sort of phantom weirdness.

**Comment by seans83** [comment]
Ok, thanks for investigating. It's frustrating because I've searched the web high and low for people who have experienced the same problem and haven't been able to find any - hence my post here. As developers start to build projects that are larger and larger and make use of many more ajax requests I find it hard to believe that others will not eventually hit the same roadblock. Is it something that could eventually be fixed? Is it simply a matter of not using PhantomJS during test discovery? Like I said, if I remove the path to the object being tested from chutzpah.json, the tests are discovered without any problems.

**Comment by mmanela** [comment]
What do you mean by "if I remove the path to the object being tested from chutzpah.json, the tests are discovered without any problem".  Isn't that the object that makes the calls?

Do these errors happen when you make ajax calls to objects that do exist?

**Comment by seans83** [comment]
Yeah that's the problem. The locations I put in those ajax calls are arbitrary, but if I called say '/<controller>/<methodname>' for any method that returns a json result in a solution, the problem still occurs. Like I said though, it only seems to happen when you go above a certain number of unit tests in  a test file. If you take half the unit tests out of that test file, and remove for example the method calls from the objects constructor - the error does not occur. 
Secondly, the error is only present during discovery - you can run those unit tests and they will all work just fine. The issue is, that because that error is reported in discovery, it's reported as an error in TFS during a build step so the test step fails, even though if the test was allowed to run regardless; it would pass. In Visual Studio, the error is reported in the output window, but you can still run the tests no problem.

Yes, it is the object that makes the calls, however, since the error is being reported during discovery and not when the tests are actually being run, it seems the problem exists because somewhere during test discovery, the test object is being parsed incorrectly. If I take the test object out of chutzpah.json, the tests get discovered anyway - so why does the object being tested need to be used during test discovery if the tests are being found anyway?

I hope that makes sense.

Sorry :)

**Comment by mmanela** [comment]
It does make sense. In order to discover tests, Chutzpah actually runs them but stubs out the bodies of the tests. It does not stub out the body of the before and after methods though which is why those run. 

So you are saying even if the Ajax methods complete it errors? That sounds like a PhantomJS bug definitely.  Could you just stub out the $.ajax method?

**Comment by seans83** [comment]
I'll try that in the repro project and see if it makes a difference.

**Comment by seans83** [comment]
If I attach the stub (e.g. spyOn('$','ajax').and.callFake(() => {})) in the beforeEach method, the error does not appear. However, the result of that spy is then being used across each test method. So if I create a test for a method that doesn't call $.ajax, the test will fail because that same spy has been used on another test method where ajax was called.

In fact, I remember trying this early on when debugging this problem. If I attach the spy/stub in each test method individually, the error is still found in discovery.

**Comment by seans83** [comment]
So I tried removing the beforeEach method entirely, and creating an instance of the test object in each test method. In each each test method, I created the stubs before creating the instance of the test object. This prevented the error from being detected. However, I then added a bunch of tests where there were no ajax/getJSON calls. I have to create the stub after I create the instance in these methods, because if you create the stub first, getJSON/ajax will be called in the test objects constructor - so all of those tests fail. The problem is, in doing this, the error returns.

So even without beforeEach, the error is still being found on discovery in each test method where the stub is not created before the test object instance. 

And yes, even if the ajax method points to a valid location, the error still occurs. And like I said, if I remove 90% of the tests in that file - there are no problems.
